### ISSUE TITLE
table inheritance 'JOINED' issue with selecting reference fields from base table

### ISSUE THREAD

**Issue by musicthebee** [doctrine/orm#5998]
I have a base task and task which is derived off of basetask using CTI

``` php
/*
 * @Entity
 * @InheritanceType("JOINED")
 * @DiscriminatorMap({"base_task" = "BaseTask",
 *     "task" = "Task"`
 * })
 */
class BaseTask 
{
 ...

   /**
     * @var User
     * @ManyToOne(targetEntity="User")
     * @JoinColumn(name="owner_id", referencedColumnName="id", onDelete="SET NULL")
     */
    protected $owner;
 ...//getter and setter
}
```

``` php
class Task extends BaseTask 
{
}
```

For any query on Task involving owner field using doctrine query builder results in this issue

```
An exception occurred while executing 'SELECT o0_.id AS id_0 FROM xyz_task w1_ INNER JOIN xyz_base_task w2_ ON w1_.id = w2_.id INNER JOIN xyz_user o0_ ON w1_.owner_id = o0_.id WHERE o0_.id = ? LIMIT 1' with params [2]:

SQLSTATE[42703]: Undefined column: 7 ERROR: column w1_.owner_id does not exist
LINE 1: ...w2_ ON w1_.id = w2_.id INNER JOIN xyz_user o0_ ON w1_.owner_...
^
HINT: Perhaps you meant to reference the column "w2_.owner_id".
```

Seems like there is an issue with walkjoin in correctly dereferencing the field name to the correct table in the inheritance hierarchy for reference fields only. It works as expected for normal fields.

Any help please?

Doctrine version is 2.5.1 and I tried 2.5.4 as well and its the same behavior.

**Comment by Ocramius** [comment]
Did you run schema validation on this? Can you turn it into a reproducible test case to be added to https://github.com/doctrine/doctrine2/tree/master/tests/Doctrine/Tests/ORM/Functional/Ticket ?

**Comment by musicthebee** [comment]
Can you tell me how to do schema validation?

I can certainly add the reproducible test case.

Thanks for quick response.

**Comment by Ocramius** [comment]
@musicthebee doctrine's CLI comes with an `orm:validate-schema` command. You can also manually use the [`SchemaValidator`](https://github.com/doctrine/doctrine2/blob/31a0c02b066764ec7b55eedc5f3cab78235a4e4b/lib/Doctrine/ORM/Tools/SchemaValidator.php), if you have a custom setup.

**Comment by musicthebee** [comment]
Okay. Will try that

**Comment by musicthebee** [comment]
`[Mapping]  OK - The mapping files are correct.
[Database] FAIL - The database schema is not in sync with the current mapping file.`

Redoing with a simple test case and will post back result

**Comment by musicthebee** [comment]
I used the `orm:schema:update` --force to update schema and that goes through fine

After that when I run `orm:validate-schema` I still get 
`[Mapping] OK - The mapping files are correct. [Database] FAIL - The database schema is not in sync with the current mapping file.`

Is there anyway to check what exactly the issue with the schema is? I tried --verbose but that doesnt give any more information

Thanks

**Comment by musicthebee** [comment]
Okay schema validation goes though fine after dropping and recreating using doctrine.

The issue is still there...

**Comment by musicthebee** [comment]
The querybuilder is built like this:

     '   return $em->getRepository('XYZ\Bundle\TaskBundle\Entity\Task')
            ->createQueryBuilder('entity')
            ->select('owner.id')
            ->innerJoin('entity.owner', 'owner')
            ->where('owner.id = :ownerId')
            ->setParameter('ownerId', $ownerId)
            ->setMaxResults(1);'

Do you see any issues with this?

Thanks

**Comment by Ocramius** [comment]
Don't see any, but if you are able to cause a failure in a test scenario, then we can at least look at it :)

**Comment by musicthebee** [comment]
i am not a great coder but I can try. Can you tell me how to run the tests? I can certainly look at an example and add

**Comment by Ocramius** [comment]
```sh
git clone git@github.com:doctrine/doctrine2.git
cd doctrine2
composer install
./vendor/bin/phpunit
```

**Comment by musicthebee** [comment]
great. Will add one...

**Comment by musicthebee** [comment]
Once done with writing test, do i need to submit a pull request?

**Comment by Ocramius** [comment]
@musicthebee ideally, yes. If you can't, making a gist is sufficient.

**Comment by musicthebee** [comment]
@Ocramius I have written the test and the it passes just fine. I am not able to get it to fail. Any ideas what I could try?

Thanks
``

Here is the test

```
<?php

namespace Doctrine\Tests\ORM\Functional;
use Doctrine\ORM\NonUniqueResultException;
use Doctrine\ORM\Query\QueryException;
use Doctrine\ORM\UnexpectedResultException;
use Doctrine\Tests\Models\CMS\CmsUser,
    Doctrine\Tests\Models\CMS\CmsArticle,
    Doctrine\Tests\Models\CMS\CmsPhonenumber;
use Doctrine\ORM\Mapping\ClassMetadata;
use Doctrine\ORM\Query;
use Doctrine\ORM\Query\Parameter;
use Doctrine\Tests\OrmFunctionalTestCase;

/**
 * Functional tests for the Class Table Inheritance mapping strategy.
 *
 * @author robo
 */
class ClassTableInheritanceTest2 extends OrmFunctionalTestCase
{
    protected function setUp()
    {
        $this->useModelSet('cms');
        parent::setUp();
            $this->_schemaTool->createSchema(array(
                $this->_em->getClassMetadata('Doctrine\Tests\ORM\Functional\CTIParent'),
                $this->_em->getClassMetadata('Doctrine\Tests\ORM\Functional\CTIGChild'),
            ));
    }

    public function testManyToOneInBase()
    {
        $user = new CmsUser;
        $user->name = 'Guilherme';
        $user->username = 'gblanco';
        $user->status = 'developer';

        $child = new CTIGChild;
        $child->setData('child');
	$child->setUser($user);

        $this->_em->persist($child);
        $this->_em->persist($user);

        $this->_em->flush();
        $this->_em->clear();

        $queryBuilder = $this->_em->getRepository('Doctrine\Tests\ORM\Functional\CTIGChild')
        			->createQueryBuilder('child')
        			->select('user.id, child.data')
        			->innerJoin('child.user', 'user')
        			->where('user.id = 1')
        			->setMaxResults(1);

        $result = $queryBuilder->getQuery()->getResult();

        $this->assertEquals(1, count($result));
    }

}

/**
 * @Entity @Table(name="cti_parents")
 * @InheritanceType("JOINED")
 * @DiscriminatorColumn(name="type", type="string")
 * @DiscriminatorMap({"parent" = "CTIParent", "gchild" = "CTIGChild"})
 */
class CTIParent {
   /**
     * @Id @Column(type="integer")
     * @GeneratedValue(strategy="AUTO")
     */
    private $id;

    /** @ManyToOne(targetEntity="Doctrine\Tests\Models\CMS\CmsUser", inversedBy="users")
     *  @JoinColumn(name="user_id", referencedColumnName="id", onDelete="SET NULL")
     */
    protected $user;

    public function getId() {
        return $this->id;
    }

    public function getUser() {
        return $this->user;
    }

    public function setUser($user) {
        $this->user = $user;
    }
}

/**
 * @Entity @Table(name="cti_gchildren")
 */
class CTIGChild extends CTIChild {

   /**
     * @Column(type="string")
     */
    protected $data;

    public function getData() {
        return $this->data;
    }

    public function setData($data) {
        $this->data = $data;
    }

}


/**
 */
class CTIChild extends CTIParent {

}
```

**Comment by musicthebee** [comment]
Okay. I was able to reproduce it. Here is the test case. It happens when you have a mappedsuperclass in the hierarchy. 

Could you please let me know if you need any additional information?

Thanks for the help


```
<?php

namespace Doctrine\Tests\ORM\Functional;
use Doctrine\ORM\NonUniqueResultException;
use Doctrine\ORM\Query\QueryException;
use Doctrine\ORM\UnexpectedResultException;
use Doctrine\Tests\Models\CMS\CmsUser,
    Doctrine\Tests\Models\CMS\CmsArticle,
    Doctrine\Tests\Models\CMS\CmsPhonenumber;
use Doctrine\ORM\Mapping\ClassMetadata;
use Doctrine\ORM\Mapping\MappedSuperClass;
use Doctrine\ORM\Query;
use Doctrine\ORM\Query\Parameter;
use Doctrine\Tests\OrmFunctionalTestCase;

/**
 * Functional tests for the Class Table Inheritance mapping strategy.
 *
 * @author robo
 */
class ClassTableInheritanceTest2 extends OrmFunctionalTestCase
{
    protected function setUp()
    {
        $this->useModelSet('cms');
        parent::setUp();
            $this->_schemaTool->createSchema(array(
                $this->_em->getClassMetadata('Doctrine\Tests\ORM\Functional\CTIParent'),
                $this->_em->getClassMetadata('Doctrine\Tests\ORM\Functional\CTIGChild'),
            ));
    }

    public function testManyToOneInBase()
    {
        $user = new CmsUser;
        $user->name = 'Guilherme';
        $user->username = 'gblanco';
        $user->status = 'developer';

        $child = new CTIGChild;
        $child->setData('child');
	$child->setUser($user);

        $this->_em->persist($child);
        $this->_em->persist($user);

        $this->_em->flush();
        $this->_em->clear();

        $queryBuilder = $this->_em->getRepository('Doctrine\Tests\ORM\Functional\CTIGChild')
        			->createQueryBuilder('child')
        			->select('user.id, child.data')
        			->innerJoin('child.user', 'user')
        			->where('user.id = 1')
        			->setMaxResults(1);

        $result = $queryBuilder->getQuery()->getResult();

        $this->assertEquals(1, count($result));
    }

}

/**
 * @Entity @Table(name="cti_parents")
 * @InheritanceType("JOINED")
 * @DiscriminatorColumn(name="type", type="string")
 * @DiscriminatorMap({"parent" = "CTIParent", "gchild" = "CTIGChild"})
 */
class CTIParent {
   /**
     * @Id @Column(type="integer")
     * @GeneratedValue(strategy="AUTO")
     */
    private $id;

    /** @ManyToOne(targetEntity="Doctrine\Tests\Models\CMS\CmsUser", inversedBy="users")
     *  @JoinColumn(name="user_id", referencedColumnName="id", onDelete="SET NULL")
     */
    protected $user;

    public function getId() {
        return $this->id;
    }

    public function getUser() {
        return $this->user;
    }

    public function setUser($user) {
        $this->user = $user;
    }
}

/**
 * @Entity @Table(name="cti_gchildren")
 */
class CTIGChild extends CTIChild {

   /**
     * @Column(type="string")
     */
    protected $data;

    public function getData() {
        return $this->data;
    }

    public function setData($data) {
        $this->data = $data;
    }

}


/**
 * @MappedSuperclass()
 */
class CTIChild extends CTIParent {

}
```

**Comment by musicthebee** [comment]
@Ocramius Would you mind confirming if this is the issue and if so, any hints on where the issue is or a patch that I can test...

Greatly appreciate your help

**Comment by musicthebee** [comment]
I have fixed it. The metadata loading doesnt take the situation where the intermediate class is mappedsuperclass. I will post the patch if there is an interest. Or send a pull request

**Comment by Ocramius** [comment]
@musicthebee thanks for looking into it! The fact that it only happens when a `mappedsuperclass` is involved is indeed pointing at a metadata loading bug. If you could send a PR, that would be awesome! I can eventually add unit tests covering what you wrote here with the larger functional test case.

**Comment by musicthebee** [comment]
okay, will do.

Thanks

**Comment by musicthebee** [comment]
> diff --git a/lib/Doctrine/ORM/Mapping/ClassMetadataFactory.php b/lib/Doctrine/ORM/Mapping/ClassMetadataFactory.php
index b42b8b5..922d66c 100644
--- a/lib/Doctrine/ORM/Mapping/ClassMetadataFactory.php
+++ b/lib/Doctrine/ORM/Mapping/ClassMetadataFactory.php
@@ -431,7 +431,9 @@ class ClassMetadataFactory extends AbstractClassMetadataFactory
                 if ($mapping['type'] & ClassMetadata::TO_MANY && !$mapping['isOwningSide']) {
                     throw MappingException::illegalToManyAssociationOnMappedSuperclass($parentClass->name, $field);
                 }
-                $mapping['sourceEntity'] = $subClass->name;
+                if (strcmp($mapping['sourceEntity'], $parentClass->name) == 0) {
+                    $mapping['sourceEntity'] = $subClass->name;
+                }
             }

             //$subclassMapping = $mapping;
[patch.zip](https://github.com/doctrine/doctrine2/files/491513/patch.zip)

**Comment by musicthebee** [comment]
Attached the patch for the fix...

**Comment by andrews05** [comment]
I think I've got the same problem, although I'm running 3.0-dev and the source of the problem appears to have shifted somewhat in the last 3 years.
I've got much the same structure as musicthebee: A JOINED inheritance type with a mappedsuperclass in the hierarchy. This is the error I get:
`Call to a member function getQuotedQualifiedName() on null in Doctrine/ORM/Persisters/Entity/JoinedSubclassPersister.php on line 504`
A quick look at the code shows it appears to be working through the hierarchy and joining to each table in the hierarchy, but the mappedsuperclass doesn't have a table which causes the error.
